### ISSUE TITLE
[core] External Redis does not work for new clusters

### ISSUE THREAD

**Issue by krfricke** [ray-project/ray#17400]
### What is the problem?

Ray 1.5.0

See https://discuss.ray.io/t/how-can-i-connect-to-pre-created-redis-instance/3048/2

Related PR:

https://github.com/ray-project/ray/pull/13170

### Reproduction (REQUIRED)
Start and run redis on Mac OS
```
brew install redis
# configure /usr/local/etc/redis.conf - e.g. bind to 0.0.0.0 and set `protected-mode no`
brew services start redis
```

Python:
```
import ray
ray.init(address="127.0.0.1:6379", _redis_password="")
```

Output after 60s:
```
RuntimeError Traceback (most recent call last)
in
----> 1 ray.init(address=“10.216.188.132:6379”, _redis_password=“password=”)

/opt/conda/envs/ACQAPython/lib/python3.6/site-packages/ray/_private/client_mode_hook.py in wrapper(*args, **kwargs)
80 if client_mode_should_convert():
81 return getattr(ray, func.name)(*args, **kwargs)
—> 82 return func(*args, **kwargs)
83
84 return wrapper

/opt/conda/envs/ACQAPython/lib/python3.6/site-packages/ray/worker.py in init(address, num_cpus, num_gpus, resources, object_store_memory, local_mode, ignore_reinit_error, include_dashboard, dashboard_host, dashboard_port, job_config, configure_logging, logging_level, logging_format, log_to_driver, namespace, runtime_env, internal_config, _enable_object_reconstruction, _redis_max_memory, _plasma_directory, _node_ip_address, _driver_object_store_memory, _memory, _redis_password, _temp_dir, _lru_evict, _metrics_export_port, _system_config, _tracing_startup_hook, **kwargs)
907 shutdown_at_exit=False,
908 spawn_reaper=False,
→ 909 connect_only=True)
910
911 if driver_mode == SCRIPT_MODE and job_config:

/opt/conda/envs/ACQAPython/lib/python3.6/site-packages/ray/node.py in init(self, ray_params, head, shutdown_at_exit, spawn_reaper, connect_only)
179 else:
180 redis_client = self.create_redis_client()
→ 181 session_name = _get_with_retry(redis_client, “session_name”)
182 self.session_name = ray._private.utils.decode(session_name)
183

/opt/conda/envs/ACQAPython/lib/python3.6/site-packages/ray/node.py in _get_with_retry(redis_client, key, num_retries)
46 time.sleep(2)
47 if not result:
—> 48 raise RuntimeError(f"Could not read ‘{key}’ from GCS (redis). "
49 “Has redis started correctly on the head node?”)
50 return result

RuntimeError: Could not read ‘session_name’ from GCS (redis). Has redis started correctly on the head node?
```

Workaround: Start ray first using cli with `ray start --head --address=127.0.0.1:6379 --redis-password ""`

cc @wuisawesome

**Comment by rkooo567** [comment]
cc @WangTaoTheTonic

**Comment by rkooo567** [comment]
Btw, do have we documented external redis stuff somewhere?

**Comment by WangTaoTheTonic** [comment]
```
----> 1 ray.init(address=“10.216.188.132:6379”, _redis_password=“password=”)
```
This log in output seems weird. I'll try to reproduce.
P.S. The doc is pretty simple: https://github.com/ray-project/ray/pull/16869/files

**Comment by CraigMeng** [comment]
Running `ray start --head --address=127.0.0.1:6379 --redis-password=password` will get following error, but I`m sure that I can connect to the redis.

`/opt/conda/envs/Ray/lib/python3.6/site-packages/ray/autoscaler/_private/cli_logger.py:61: FutureWarning: Not all Ray CLI dependencies were found. In Ray 1.4+, the Ray CLI, autoscaler, and dashboard will only be usable via pip install 'ray[default]'. Please update your install command.
“update your install command.”, FutureWarning)
Will use value of --address as remote Redis server address(es). If the primary one is not reachable, we starts new one(s) with --port in local.
Local node IP: 10.240.16.215
2021-07-29 01:47:53,602 ERROR services.py:1254 – Failed to start the dashboard: Failed to start the dashboard, return code 1. The last 10 lines of /tmp/ray/session_2021-07-29_01-47-51_520908_774/logs/dashboard.log:
2021-07-29 01:47:53,398 INFO dashboard.py:92 – Setup static dir for dashboard: /opt/conda/envs/Ray/lib/python3.6/site-packages/ray/new_dashboard/client/build
2021-07-29 01:47:53,432 ERROR head.py:87 – Connect to GCS failed: GCS address not found., retry…

2021-07-29 01:47:53,618 WARNING services.py:1716 – WARNING: The object store is using /tmp instead of /dev/shm because /dev/shm has only 67108864 bytes available. This will harm performance! You may be able to free up space by deleting files in /dev/shm. If you are inside a Docker container, you can increase /dev/shm size by passing ‘–shm-size=2.40gb’ to ‘docker run’ (or add it to the run_options list in a Ray cluster config). Make sure to set this to more than 30% of available RAM.
[2021-07-29 01:47:53,795 C 774 774] redis_client.cc:126: Check failed: _s.ok() Bad status: RedisError:
*** StackTrace Information ***
ray::SpdLogMessage::Flush()
ray::RayLog::~RayLog()
ray::gcs::RedisClient::Connect()
ray::gcs::RedisClient::Connect()
ray::gcs::ServiceBasedGcsClient::Connect()
ray::gcs::GlobalStateAccessor::Connect()
__pyx_pw_3ray_7_raylet_19GlobalStateAccessor_3connect()
_PyCFunction_FastCallDict`

**Comment by krfricke** [comment]
Can you try with `--external-address`?

**Comment by krfricke** [comment]
By the way @rkooo567 @WangTaoTheTonic the conflicting block is here: https://github.com/ray-project/ray/blob/master/python/ray/node.py#L179-L182

If an `address` is given in `ray.init()` it is always assumed that `head=False`

**Comment by CraigMeng** [comment]
> Can you try with `--external-address`?

Using ray==1.5.0, replacing --address with --external-address will get following error:

ray start --head --external-address="{host}.redis.cache.windows.net:6379" --redis-password="password="
Usage: ray start [OPTIONS]
Try 'ray start --help' for help.

Error: No such option: --external-address (Possible options: --address, --node-ip-address)

**Comment by rkooo567** [comment]
Hmm we should add a test for this feature. @WangTaoTheTonic is it possible to make a PR to test Ray with external Redis?

**Comment by CraigMeng** [comment]
Hi，I‘d love to help on this feature，and I have a proposal. 

So firstly I think we need to discuss my proposal. where should I begin with？

@WangTaoTheTonic @krfricke @rkooo567

**Comment by WangTaoTheTonic** [comment]
Hi @CraigMeng , we did not support external redis from ray.init but only `ray start`, but I think we need to do that.
Glad to hear your idea! If it's simple you can just write it here or create a pull request for that.

**Comment by wuisawesome** [comment]
@WangTaoTheTonic do you know if the current implementation supports starting multiple ray clusters with the same redis server?

**Comment by WangTaoTheTonic** [comment]
> @WangTaoTheTonic do you know if the current implementation supports starting multiple ray clusters with the same redis server?

No we can't start up multiple ray clusters on one same redis server. It is not supported. Neither with external redis or internal redis in the head node.

**Comment by WangTaoTheTonic** [comment]
Hi @krfricke , I took a look at the codes and it seems like that the rules are:
-  if we use `ray.init` with passing a `address`, it means we will connected an existing cluster
- if we use `ray.init` without `address` parameter, it will start a new one locally.
So the external redis support is nothing to do with `ray.init`. It just is used in `ray start` command to connect an existing redis server when we start a new ray cluster.

Hi @CraigMeng , I installed the ray 1.5.0 and started a new ray cluster with an existing redis server, and the cluster will be started up. (but only with an dashboard error which I will try to fix it)
![image](https://user-images.githubusercontent.com/5276001/129470926-fb922043-e9f9-4723-9714-96ba85796d82.png)
Can you make sure the `--address` and `--redis-password` is right? (I guess the redis password should be empty string if you start redis server without setting the `requirepass `

**Comment by WangTaoTheTonic** [comment]
After another try I figured out that dashboard cann't start up if the redis server requires no passworld it use `aioredis` to get its redis client, in which it will use `AUTH` command and it will response with an error if we pass empty string as password.
![image](https://user-images.githubusercontent.com/5276001/129473503-015407ef-dc08-4292-b335-b364ed60d7cb.png)
![image](https://user-images.githubusercontent.com/5276001/129473515-8e35c6f2-7cd6-4b80-9529-65bcd08d5d18.png)
So we must use external redis with specified `requirepass`.

**Comment by CraigMeng** [comment]
@WangTaoTheTonic Thank you, I can run the cluster now. Perhaps I give the wrong password by mistake.

**Comment by CraigMeng** [comment]
@WangTaoTheTonic So I states my thoughts here:
we may distinguish different mode by specifying protocol. 

Currently we are able to use ray.init(address="ray://123.45.67.89:10001") to connect to existing ray cluster. And I know that Ray has the plan to remove Redis as hard dependency (https://github.com/ray-project/ray/issues/7239). In the future, perhaps we can use other backend to support GCS.

So the plan is:

1.  if user passes in {IP | HOST}:{PORT} or ray://{IP | HOST}:{PORT}, we go exactly the current way

2. if user passes redis://{IP | HOST}:{PORT}, then we know this is an external Redis and we need to connect to that instance instead of creating new one. We can even support TLS protected Redis instance by specifying rediss://{IP | HOST}:{PORT}. This is helpful when external Redis instance comes from cloud provider, such as Azure Cache for Redis.

3. In the future, if Ray supports more pluggable backend for GCS, we can use other prefix protocol.

**Comment by WangTaoTheTonic** [comment]
@CraigMeng Thanks for your reply and ideas!
The current way of using address in `Ray.init` is:
1. `ray.init()` with no address provided will start ray processes including redis in local and connect to it.
2. `ray.init("auto")` will try to connect to port 6379 in local.
3. `ray.init("ip:port")` will try to connect to the port on the ip specified.
4. `ray.init("ray://ip:port")` will try to connect to the ray client server proxy, used by ray client feature.

Adding prefix of `redis://` seems a nice idea to indicate that we are trying to connect a cluster whose backend is redis but i think it's unnecessary as user has no need to know what type of backend storage is.

For removing redis dependency and supporting pluggable backend I think it is a long term work and let's see if there's better ideas then. Is it ok?

**Comment by CraigMeng** [comment]
@WangTaoTheTonic I assume item3 and item4 will do the same thing according to documentation (https://docs.ray.io/en/master/package-ref.html?#ray.init)? Correct me if I am wrong.

Yeah I agree that adding prefix exposes underlying implementation details to users. I thought ray.init will do exactly same thing with ray start, so I was trying to resolve the conflicts that how ray.init handles passing in Redis address.

By re-reading the docs, I finally find my self wrong - ray.init will never create a ray cluster but only processes. This implementation is fine to me, then. 

Thank a lot

**Comment by WangTaoTheTonic** [comment]
@CraigMeng Thanks for the sharing. You're alway welcome. :-)
@krfricke Hi Kai, does the explanation sound good to you? We'll close this if it's alright.

**Comment by krfricke** [comment]
Yes, this is great. Thanks @WangTaoTheTonic!

**Comment by tisonkun** [comment]
It seems that `--address` doesn't work if specified as an existing redis server address:

```
$ ray start --head --address=127.0.0.1:6379 --redis-password=""
Usage stats collection will be enabled by default in the next release. See https://github.com/ray-project/ray/issues/20857 for more details.
Specifying --address for external Redis address is deprecated. Please specify environment variable RAY_REDIS_ADDRESS=127.0.0.1:6379 instead.
Will use `127.0.0.1:6379` as external Redis server address(es). If the primary one is not reachable, we starts new one(s) with `--port` in local.
Local node IP: 127.0.0.1
2022-05-09 15:43:34,212	WARNING utils.py:1248 -- Unable to connect to GCS at 127.0.0.1:6379. Check that (1) Ray GCS with matching version started successfully at the specified address, and (2) there is no firewall setting preventing access.
^C
Aborted!
```

**Comment by WangTaoTheTonic** [comment]
> It seems that `--address` doesn't work if specified as an existing redis server address:
> 
> ```
> $ ray start --head --address=127.0.0.1:6379 --redis-password=""
> Usage stats collection will be enabled by default in the next release. See https://github.com/ray-project/ray/issues/20857 for more details.
> Specifying --address for external Redis address is deprecated. Please specify environment variable RAY_REDIS_ADDRESS=127.0.0.1:6379 instead.
> Will use `127.0.0.1:6379` as external Redis server address(es). If the primary one is not reachable, we starts new one(s) with `--port` in local.
> Local node IP: 127.0.0.1
> 2022-05-09 15:43:34,212	WARNING utils.py:1248 -- Unable to connect to GCS at 127.0.0.1:6379. Check that (1) Ray GCS with matching version started successfully at the specified address, and (2) there is no firewall setting preventing access.
> ^C
> Aborted!
> ```
> 
> If I change the address to a spare port, it will start a ray instance with GCS runs on that port, though.

Which release did you use? If it's not 1.5.0 please file another new issue for that and copy to me. I'm glad to take a look.

**Comment by tisonkun** [comment]
@WangTaoTheTonic IIRC this issue is resolved in the WeChat group, while #24634 created for a ticket improving document.
